### DISCUSSION TITLE
2. Transaction

### DISCUSSION THREAD

**Discussion by 24tngus** [24tngus/CS_STUDY#15] [Database]
#### 1. 트랜잭션 정의 

- (사용자 입장) 작업의 논리적 단위, (시스템 입장) 데이터를 접근 또는 변경하는 프로그램의 단위
- DB에서 논리적 기능을 수행하고, DB 상태를 변화시키는 작업의 단위 
- 질의어(SQL)를 이용하여 데이터베이스에 접근하는 것 (여러 개의 쿼리를 하나로 묶는 단위)
- 기능
  - 작업의 `완전성` 보장
  - 완벽하게 처리하지 못할 경우 원 상태로 복구해서 작업의 일부만 적용되지 않게 만드는 기능
  
> 📖 참고 📖 Query 
- ✏️ 정의
  - DB에 특정 데이터를 요청하는 일 
  - query = 질의문 = SQL
- ✏️ 종류
  - SELECT
  - DELETE
  - UPDATE

> 📖 참고 📖 트랜잭션과 Lock 차이
- ✏️ Transaction : 데이터 정합성을 보장하는 기능
  - 하나 혹은 여러개의 변경 작업을 수행하는 쿼리에 관계없이 
  - 논리적인 작업이 100% 적용되거나 아무것도 적용되지 않도록 보장함
- ✏️ Lock : 동시성 제어하는 기능
  - 여러 커넥션에서 동시에 동일한 자원(레코드/테이블)을 요청할 경우
  - 순서대로 한 커넥션만 변경하게 해줌 

#### 2. commit과 rollback

- 데이터 무결성 보장 및 데이터 변경 사항을 쉽게 확인 가능 (작업 그룹화)
- 트랜잭션으로 묶는 로직에 외부 API호출 자제 

- commit 
  - 여러 쿼리가 성공적으로 처리되었다고 확정하는 명령어
  - 트랜잭션 단위(update, insert, delete 등)로 수행
  - 변경된 내용이 DB에 영구적으로 저장됨
  - 커밋 수행 = 트랜잭션 성공적으로 수행

- rollback
  - 트랜잭션 처리 전 상황으로 복귀(취소)하는 일
  - 에러 및 이슈 발생시 이전으로 돌림

#### 3. [트랜잭션 전파 (Transaction Propagation)](https://mangkyu.tistory.com/269)

- 트랜잭션 시작
  - 트랜잭션 수행시 Connection 객체를 가져와 사용
  - `setAutoCommit(false)`로 트랜잭션 시작을 선언 (JDBC 기본 설정으로 DB작업 수행 직후 자동 커밋 옵션 활성화 되어있어서 false로 설정함)

- 트랜잭션 종료
  - 시작되면 `commit()` 또는 `rollback()` 호출될 때까지 하나의 트랜잭션

- 트랜잭션 전파
  - 트랜잭션이 진행중일 때 추가 트랜잭션 진행을 어떻게 할지 결정하는 것 
  - 여러 트랜잭션을 하나의 트랜잭션으로 묶음
  - 기존 트랜잭션에 참여 / 별도의 트랜잭션으로 진행 / 에러 발생시킴
  
  - Spring 프레임워크에서 `@Transactional` 애너테이션을 통해 여러 쿼리 코드를 1개의 트랜잭션으로 처리함
  ```java
  @Service
  @Transactional(readOnly = true)
  public class MemberService{
  	private final MemberRepository memberRepository;
    
    public MemberService(MemberRepository memberRepository) {
    	this.memberRepository = memberRepository;
     }
  }
  ```
  - 트랜잭션 전파 사용시 논리 트랜잭션 개념 사용됨
    ![](https://velog.velcdn.com/images/24tngus/post/fc8aa14e-8bcb-4a1d-818e-5d3460db82c0/image.png) 
  - 물리 트랜잭션 : 실제 데이터베이스에 적용되는 트랜잭션 (Connection을 통해 commit/rollback하는 단위)
  - 논리 트랜잭션 : 스프링이 트랜잭션 매니저를 통해 트랜잭션을 처리하는 단위 

- 🗒️ 예시 : 게시판에 글 작성하는 기능
  - 글을 임시 저장하는 기능(A)이 글 작성 중간에 수행되고, 동시에 로깅(B)을 해야하는 상황
  - 만약, 로깅(B) 작업에서 예외 발생될 경우, 트랜잭션 전파 옵션이 Defaul이면 로깅(B)이 Rollback 진행되면서 임시 저장 기능(A)도 함께 Rollback 발생
  - 하지만, 위의 경우 A가 Rollback 되면 안되기 때문에, 트랜잭션 전파 속성 설정 필요함

> 📖 참고 📖 트랜잭션 전파 속성
- REQUIRED
  - 트랜잭션이 필요하고, 없으면 새로 만듦 (Default 속성)
  - 기존 트랜잭션 없음 : 새로운 트랜잭션 생성함
  - 기존 트랜잭션 있음 : 기존 트랜잭션에 참여함
  - 기본적으로 논리 트랜잭션 2개를 묶어 1개의 물리 트랜잭션을 사용함
    - 모든 논리 트랜잭션이 커밋되어야 물리 트랜잭션 커밋됨
    - 하나의 논리 트랜잭션이라도 롤백되면 물리 트랜잭션 롤백됨
  - 트랜잭션 매니저에 의해 관리되는 논리 트랜잭션이 존재하므로 커밋 총 2회(내부 1회, 외부 1회) 실행
  ![](https://velog.velcdn.com/images/24tngus/post/fc8aa14e-8bcb-4a1d-818e-5d3460db82c0/image.png) 
- REQUIRES_NEW
  - 항상 새로운 트랜잭션이 필요함
  - 기존 트랜잭션 없음 : 새로운 트랜잭션 생성
  - 기존 트랜잭션 있음 : 기존 트랜잭션을 보류시키고 새로운 트랜잭션 생성함
  - 외부/내부 트랜잭션을 완전히 분리하는 전파 속성 (내부 트랜잭션 롤백이 외부 트랜잭션 롤백에 영향 X)
  - 각각 트랜잭션 별로 커밋과 롤백이 수행됨 (각각의 DB Connection 사용됨, 1개의 HTTP 요청에 대해 2개의 Connection 사용)
  - 내부 트랜잭션이 처리 중일 경우, 외부 트랜잭션이 대기함 (DB Connection 고갈 주의)
      ![](https://velog.velcdn.com/images/24tngus/post/6208ce59-084d-4e31-bfeb-cae01f18c566/image.png)
- SUPPORTS
  - 트랜잭션이 있으면 지원함 (트랜잭션이 없어도 됨)
  - 기존 트랜잭션 없음 : 트랜잭션 없이 진행함
  - 기존 트랜잭션 있음 : 기존 트랜잭션에 참여함
- MANDATORY
  - 트랜잭션이 의무임 (트랜잭션 반드시 필요함)
  - 기존 트랜잭션 없음 : illegalTransactionStateException 예외 발생
  - 기존 트랜잭션 있음 : 기존 트랜잭션에 참여함
- NOT_SUPPORTED
  - 트랜잭션을 지원하지 않음 (트랜잭션 없이 진행함)
  - 기존 트랜잭션 없음 : 트랜잭션 없이 진행함
  - 기존 트랜잭션 있음 : 기존 트랜잭션을 보류시키고 트랜잭션 없이 진행함
- NEVER
  - 트랜잭션을 사용하지 않음 (기존 트랜잭션도 허용하지 X)
  - 기존 트랜잭션 없음 : 트랜잭션 없이 진행
  - 기존 트랜잭션 있음 : illegalTransactionStateException 예외 발생
- NESTED
  - 중첩(자식) 트랜잭션 생성함
  - 기존 트랜잭션 없음 : 새로운 트랜잭션을 생성함
  - 기존 트랜잭션 있음 : 중첩 트랜잭션을 만듦
  - 독립적인 트랜잭션(REQUIRES_NEW)이 아닌 이미 진행중인 트랜잭션에 중첩(자식) 트랜잭션 생성
    - 부모 트랜잭션의 영향(커밋과 롤백)은 받지만, 외부에 영향을 주지 않음
    - 중첩 트랜잭션이 롤백이 되어도 외부 트랜잭션은 커밋 가능하지만, 외부 트랜잭션이 롤백될 경우 중첩 트랜잭션도 함께 롤백됨
  
#### 4. 트랜잭션 특징 (ACID)

- (1) 원자성 atomicity
  - 트랜잭션이 DB에 모두 반영되거나, 전혀 반영되지 않음 (All or Nothing)
  - 트랜잭션 중간에 문제 발생할 경우, 트랜잭션에 해당하는 어떠한 작업도 수행되지 않고 아무런 문제가 발생하지 않았을 경우에만 모든 작업 수행함
- (2) 일관성 consistency
  - 트랜잭션 작업 처리의 결과가 항상 동일함 (허용된 방식으로만 데이터 변경 가능)
  - 트랜잭션이 일어나기 전의 상황과 동일하게 데이터의 일관성 보장
- (3) 격리성 isolation
  - 하나의 트랜잭션은 다른 트랜잭션에 끼어들 수 없음 (간섭 불가능)
  - 각각의 트랜잭션은 서로의 간섭없이 독립적으로 수행
- (4) 지속성 durability
  - 트랜잭션이 성공적으로 완료되면 영구적으로 결과에 저장됨 (commit)
  
> 📖 참고 📖 무결성
- 데이터의 정확성, 일관성, 유효성을 유지하는 것

#### 5. 트랜잭션의 격리 수준

- 트랜잭션이 순차적으로 실행될 경우 : 격리성⬆️, 동시성⬇️, 성능⬇️ 
- 격리 수준 단계
  - Serializable : 이상현상 발생 X (격리성⬆️)
  - Repatable Read : 팬텀 리드 발생 가능
  - Read Commited : + 반복 가능하지 않은 조회 발생 가능
  - Read Uncommited : + 더티 리드 발생 가능 (격리성⬇️)

| Isolation Level|Dirty reads|Non-repeatable reads|Phantoms   |
|----------------|:---------:|:------------------:|:---------:|
|Read Uncommited |May occur  |May occur           |May occur  |
|Read Commited   |Don't occur|May occur           |May occur  |
|Repeatable Read |Don't occur|Don't occur         |May occur  |
|Serializable    |Don't occur|Don't occur         |Don't occur|
 
 ![](https://velog.velcdn.com/images/24tngus/post/f7f1e20c-5506-486b-b7e9-f3a9bde72552/image.png)

- 격리 수준 설정
```java
set session transaction isolation level read uncommitted;
```

- Serializable
  - 커밋 완료된 데이터만 조회 가능
  - 트랜잭션을 순차적으로 진행 (간섭 발생 X)
  - 여러 트랜잭션이 동시에 같은 행에 접근 불가

- Repatable Read
  - 커밋 완료된 데이터만 조회 가능
  - 같은 행을 반복해서 조회하더라도 동일한 결과 보장
  - 한 트랜잭션에서 수정한 행을 다른 트랜잭션이 수정 불가 / 새로운 행 추가 가능 (➡️ 팬텀 리드 발생)
  - MySQL8.0의 innoDB의 기본값
  
- Read Commited 
  - 커밋 완료된 데이터만 조회 가능
  - 같은 행을 반복해서 조회하더라도 동일한 결과 보장X
  - 한 트랜잭션에서 수정한 행을 다른 트랜잭션이 수정 + 새로운 행 추가 가능 (➡️ 팬텀 리드, 반복 가능하지 않은 조회 발생)
  - PostgreSQL, SQL Server, 오라클의 기본값
  
- Read Uncommited 
  - 커밋하지 않은 데이터 조회 가능
  - 속도 가장 빠름
  
#### 6. 트랜잭션 이상 현상

- (1) 팬텀 리드(Phantom reads)
  - 한 트랜잭션 내에서 동일한 쿼리를 2번 이상 보냈을 때 (다른 트랜잭션의 영향으로) 조회 결과가 다름
  - 🗒️ 예시 : 새로운 레코드 추가 및 삭제되어 발생
  ![](https://velog.velcdn.com/images/24tngus/post/b93ca3bb-da1f-44b8-8a8b-7a58d407571b/image.png)
  
- (2) 반복 가능하지 않은 조회 (Nonrepatable Reads)
  - 한 트랜잭션 내의 같은 행에 2번 이상 조회가 발생 했을 때, 그 값이 다른 것
  - 🗒️ 예시 : 트랜잭션 A는 SELECT하는 구문이고, 트랜잭션 B는 데이터 변경하는 구문일 경우 트랜잭션 A는 B때문에 반복 수행시 다른 결과 가져옴
    ![](https://velog.velcdn.com/images/24tngus/post/d52a17d1-9adc-4a44-bdbe-7ada2b00159b/image.png)

- (3) 더티 리드 (Dirty Reads)
  - 한 트랜잭션이 다른 트랜잭션의 아직 커밋되지 않은 데이터를 읽음
  - 🗒️ 예시 : 트랜잭션 A가 데이터를 변경하고, 커밋하기 전에 트랜잭션 B가 데이터를 읽을 경우 잘못된 데이터 읽음
  ![](https://velog.velcdn.com/images/24tngus/post/5105c6d0-5fc7-4d93-8b30-455c9e8a799f/image.png)

> 📖 참고 📖 트랜잭션 사용시 주의
- 트랜잭션은 꼭 필요한 최소한 코드에만 적용하기
- 일반적으로 DB Connection 개수는 제한적이므로, 각 단위 프로그램에서 Connection 소유 시간이 길어지면 사용 가능한 여유 Connection 개수가 줄어듦
- 따라서, 트랜잭션 범위 최소화

**Comment by berryberries** [comment]
확인했습니다.

**Comment by jindream6128** [comment]
멘토님 Comment) 
격리 수준은 성능과도 연관이 있는데 같이 알아보면 더욱 좋을 듯 합니다. (격리 수준이 높을 수록 동시성이 저하되고 성능이 낮아짐) 그리고 격리 수준의 경우 DB 벤더마다 살짝식 다른 특성이 있는 것도 보시면 좋을 듯 합니다. (InnoDB의 경우 팬텀 리드 현상이 없음 등)

**Comment by jindream6128** [comment]
확인하였습니다!
멘토님 말에있는 DB 엔진마다 격리수준이 다르다면 큰 차이가 있는지 확인하는 것도 좋을 것 같아요 !

**Comment by sujin-park0607** [comment]
확인했습니다!

**Comment by solo5star** [comment]
원자성(Atomicity)는 All or Nothing 이라고도 자주 언급되는 것 같습니다. 전부 성공하거나, 실패하기 때문이에요.

트랜잭션이 교착상태(Deadlock)을 피하기 위해 어떻게 락을 거는지도 한번 보시면 좋을 것 같아요! (2PL, Two-Phase Locking)

물론 2PL을 적용하여도 여전히 트랜잭션은 교착상태에 빠질 수 있고 실패할 수 있다는 점을 알아두시면 좋을 것 같습니다.
